src/slack/metricsNewsletter.ts

```diff
@@ -0,0 +1,153 @@
+} from "~/src/data/metrics";
+import { bullet, section, type TextElement } from "~/src/slack/blocks";
+
+type SlackBuildingBlock = KnownBlock | Block;
+
+function countMetricsByTrend(metrics: Metric[]): {
+  better: number;
+  worseOrSame: number;
+} {
+  return metrics.reduce(
+    (acc, metric) => {
+      if (!metric.previousSample) {
+        return { ...acc, worseOrSame: acc.worseOrSame + 1 };
+      }
+
+      const trend = calculateTrend([
+        metric.previousSample.value,
+        metric.currentSample.value,
+      ]);
+      const isBetter =
+        getTrendDirection(trend, metric.definition.positiveDirection) ===
+        "better";
+      return {
+        better: acc.better + Number(isBetter),
+        worseOrSame: acc.worseOrSame + Number(!isBetter),
+      };
+    },
+    { better: 0, worseOrSame: 0 },
+  );
+}
+
+function formatDate(
+  metric: Metric,
+  sample: "currentSample" | "previousSample",
+): string {
+  const date =
+    metric.definition.type === "band"
+      ? // `label` is in arbitrary format, so we can't really parse it and format it.
+        metric[sample]!.label ?? "Chybějící datum"
+      : Intl.DateTimeFormat("cs-CZ", { day: "numeric", month: "long" }).format(
+          new Date(metric[sample]!.date),
+        );
+  return date.replace(/ /g, "\u00A0");
+}
+
+/**
+ * Convert a metric sample to a list of `TextElement` objects.
+ * Used in Slack Building Blocks.
+ */
+function sampleAsElements(
+  metric: Metric,
+  sample: "currentSample" | "previousSample",
+): TextElement[] {
+  return [
+    { text: `${metric[sample]!.value}`, style: { italic: true } },
+    { text: `\u00A0(${formatDate(metric, sample)})` },
+  ];
+}
+
+function prepareMetricBlock(metric: Metric): SlackBuildingBlock {
+  const sectionElements: TextElement[] = [
+    { text: metric.definition.qualifiedName, style: { bold: true } },
+  ];
+  const bulletElements: TextElement[] = [];
+
+  if (metric.previousSample) {
+    const trend = calculateTrend([
+      metric.previousSample.value,
+      metric.currentSample.value,
+    ]);
+    const isTrendBetter =
+      getTrendDirection(trend, metric.definition.positiveDirection) ===
+      "better";
+    const emoji = isTrendBetter ? "🟢" : "🔴";
+    // Section text if there is a previous sample
+    sectionElements.push({
+      text: ` ${getTrendIcon(trend)}\u00A0${trend}\u00A0%\u00A0${emoji}`,
+    });
+
+    // Bullet text if there is a previous sample
+    bulletElements.push(...sampleAsElements(metric, "previousSample"), {
+      text: " → ",
+    });
+  }
+
+  bulletElements.push(...sampleAsElements(metric, "currentSample"));
+  return {
+    type: "rich_text",
+    elements: [section(sectionElements), bullet(bulletElements)],
+  };
+}
+
+/**
+ * Return the metrics newsletter header, formatted via Slack's building blocks.
+ *
+ * For more info about building blocks, see: https://api.slack.com/block-kit/building
+ */
+export async function getMetricsNewsletterHeader(): Promise<
```

Comment:
Ty máš ty funkce napsané tak, že přímo v sobě volají to čtení dat z API. IMO je lepší to dělat tak, že převezmou potřebná data jako parametr. Je to testovatelnější a ty funkce se pak dají líp použít v trochu jiném kontextu, takže to celkově líp snáší změny. (Tady to nevidím jako zásadní, záleží na tobě, jestli se ti to chce předělat.)

Reply:
Souhlasim s tebou. Děkuji za podnět!